File: AlienDictionary.java
/*
* Did this code successfully run on Leetcode : YES
* 
* Any problem you faced while coding this : NO
* 
* Time Complexity: O(n*k + 26 characters in queue)
    n = total words
    k = avg length of each word
* 
* Space Complexity: O(1)
    hashmap with 26 character keys and indegrees array with 26 elements.
* 
*/

import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class AlienDictionary {
    class Solution {
        int[] indegrees;

        HashMap<Character, List<Character>> adjList;

        private void populateAdjacencyList(String[] words) {
            // populating the adj list keys because we are stopping whenever there is a
            // breach
            // this might affect the indegree calculation while applying topological sort
            for (int index = 0; index < words.length; index++) {
                for (int i = 0; i < words[index].length(); i++) {
                    char ch = words[index].charAt(i);

                    if (!adjList.containsKey(ch)) {
                        adjList.put(ch, new ArrayList<>());
                    }
                }
            }

            for (int index = 0; index < words.length - 1; index++) {
                String word1 = words[index];

                String word2 = words[index + 1];

                // apple and app -> invalid order, so clear the map with added charactes
                // so that queue dont find any indegree chars and return ""
                if (word1.startsWith(word2) && word1.length() > word2.length()) {
                    adjList.clear();
                    return;
                }

                for (int wordIndex = 0; wordIndex < word1.length() && wordIndex < word2.length(); wordIndex++) {
                    char word1Char = word1.charAt(wordIndex);

                    char word2Char = word2.charAt(wordIndex);

                    // breach
                    if (word1Char != word2Char) {
                        char preceedingChar = word1Char;

                        char nextChar = word2Char;

                        adjList.get(preceedingChar).add(nextChar);

                        indegrees[nextChar - 'a']++;

                        break;
                    }
                }
            }
        }

        public String alienOrder(String[] words) {
            indegrees = new int[26];

            adjList = new HashMap<>();

            populateAdjacencyList(words);

            Queue<Character> queue = new LinkedList<>();

            StringBuilder str = new StringBuilder();

            for (char key : adjList.keySet()) {
                if (indegrees[key - 'a'] == 0) {
                    queue.add(key);
                    str.append(key);
                }
            }

            while (!queue.isEmpty()) {
                char ch = queue.poll();

                // process edges of ch
                for (char node : adjList.get(ch)) {
                    indegrees[node - 'a']--;

                    if (indegrees[node - 'a'] == 0) {
                        queue.add(node);
                        str.append(node);
                    }
                }
            }

            if (str.length() == adjList.size()) {
                return str.toString();
            }

            return "";
        }
    }
}

File: Alien Dictionary.py
""""// Time Complexity : O(n.k); where n is number of words and k is avg length of a word.
// Space Complexity : O(1) as map will at most have 26 chars
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :
"""
from collections import deque
class Solution:
    def alienDictionary(self, words):
        h={}
        indegree=[0]*26

        n=len(words)
        # print(n)
        for i in range(n):
            word=words[i]
            for j in range(len(word)):
                if word[j] not in h:
                    h[word[j]]=set()
        for i in range(n-1):
            first=words[i]
            second=words[i+1]
            fl=len(first)
            sl=len(second)

            if fl>sl and first[:sl]==second:
                h=[]
                break

            p1=0
            p2=0

            while p1<fl and p2<sl:
                out=first[p1]
                inn =second[p2]
                if out!=inn:
                    sets=h[out]
                    if inn not in sets:
                        indegree[ord(inn)-ord('a')]+=1
                        sets.add(inn)
                    break
                p1+=1
                p2+=1

        q=deque()
        result=[]

        for c in h:
            if indegree[ord(c)-ord('a')]==0:
                q.append(c)
                result.append(c)

        if len(q)==0:
            return ""
        while q:
            c=q.popleft()
            edges=h[c]
            for edge in edges:
                indegree[ord(edge)-ord('a')]-=1
                if indegree[ord(edge)-ord('a')]==0:
                    q.append(edge)
                    result.append(edge)

        if len(result)!=len(h):
            return ""

        return ''.join(result)

Obj=Solution()
print(Obj.alienDictionary(["abc", "ab"]))
# print(Obj.alienDictionary(["wrt", "wrf", "er", "ett", "rftt"]))








